contracts/Core/BlockManager.sol

```diff
@@ -120,31 +120,30 @@ contract BlockManager is Initializable, ACL, BlockStorage, StateManager, IBlockM
         require(stakerId > 0, "Structs.Staker does not exist");
         require(blocks[epoch].proposerId == 0, "Block already confirmed");
 
-        if (sortedProposedBlockIds[epoch].length == 0 || blockIndexToBeConfirmed == -1) return;
-
-        uint8 blockId = sortedProposedBlockIds[epoch][uint8(blockIndexToBeConfirmed)];
-        uint32 proposerId = proposedBlocks[epoch][blockId].proposerId;
+        uint8[] memory deactivatedAssets = assetManager.getPendingDeactivations();
+        if (sortedProposedBlockIds[epoch].length == 0 || blockIndexToBeConfirmed == -1) {
+            for (uint8 i = 0; i < deactivatedAssets.length; i++) {
+                // slither-disable-next-line calls-loop
+                assetManager.deactivateCollection(epoch, deactivatedAssets[i]);
+            }
+            return;
+        }
+        uint32 proposerId = proposedBlocks[epoch][sortedProposedBlockIds[epoch][uint8(blockIndexToBeConfirmed)]].proposerId;
         require(proposerId == stakerId, "Block can be confirmed by proposer of the block");
-        emit BlockConfirmed(epoch, proposerId, proposedBlocks[epoch][blockId].medians, block.timestamp);
-        blocks[epoch] = proposedBlocks[epoch][blockId];
-        rewardManager.giveBlockReward(stakerId, epoch);
-        randomNoProvider.provideSecret(epoch, voteManager.getRandaoHash());
+        _confirmBlock(epoch, deactivatedAssets, proposerId);
     }
 
     function confirmPreviousEpochBlock(uint32 stakerId) external override initialized onlyRole(BLOCK_CONFIRMER_ROLE) {
         uint32 epoch = parameters.getEpoch();
-        if (sortedProposedBlockIds[epoch - 1].length == 0 || blockIndexToBeConfirmed == -1) return;
-        uint8 blockId = sortedProposedBlockIds[epoch - 1][uint8(blockIndexToBeConfirmed)];
-        blocks[epoch - 1] = proposedBlocks[epoch - 1][blockId];
-
-        emit BlockConfirmed(
-            epoch - 1,
-            proposedBlocks[epoch - 1][blockId].proposerId,
-            proposedBlocks[epoch - 1][blockId].medians,
-            block.timestamp
-        );
-        rewardManager.giveBlockReward(stakerId, epoch - 1);
-        randomNoProvider.provideSecret(epoch - 1, voteManager.getRandaoHash());
+        uint8[] memory deactivatedAssets = assetManager.getPendingDeactivations();
+        if (sortedProposedBlockIds[epoch - 1].length == 0 || blockIndexToBeConfirmed == -1) {
+            for (uint8 i = 0; i < deactivatedAssets.length; i++) {
```

Comment:
Thanks for pointing out that the conditions in both the for loops are different, need to change that.
Coming to your concern, the first for loop will run only if the condition is satisfied and we return in L144 like we used to do earlier when there wasnt any valid block

https://github.com/razor-network/contracts/blob/8c86f3a9dd035d5cc233a4fe90adb24c201b1535/contracts/Core/BlockManager.sol#L144

This was done to ensure that collections are deactivated regardless of whether there is a valid block or no. Either one of the for loops will be executed, never both. If there is a valid block, we do not run the first loop and go to logic of confirming the block and the second for loop

Reply:
hmm cool got it
maybe as this deactivateCollection is callable by blockManager only 
and assetManager already knows pendingDeactivations 

can't we refractor deactivateCollection such that
call to it will only act as trigger to deactivate all pending collections lies in assetManager.
hence for loop will be there.

There are three advantages of doing this 
There will be only one external call, rather than o(n) we have now 
Fetching pendingCollections and again feeding it to AssetManager only don't make sense, assetManager already have it
There will be single for loop, so we save space